core/domain/story_services.py

```diff
@@ -141,6 +155,20 @@ def apply_change_list(story_id, change_list):
                       story_domain.STORY_NODE_PROPERTY_EXPLORATION_ID):
                     story.update_node_exploration_id(
                         change.node_id, change.new_value)
+                    if change.old_value:
+                        if change.old_value in exp_ids_added_to_story:
```

Comment:
While parsing through this logic, another solution occurred to me.

If we are just maintaining these lists so that we can update the context models, why don't we just calculate them outside this method by comparing the old story chapter exp IDs and the new ones? The diff can be determined from that, rather than doing a play-by-play of all the operations. This seems feasible especially since we don't actually need to keep track of the order of operations in order to figure out what to do with the context models...

I think that might make the logic simpler? Sorry for not thinking of this before.

Reply:
Done